cmd/annotation.go

```diff
@@ -0,0 +1,161 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/deis/pkg/prettyprint"
+	"regexp"
+
+	"github.com/deis/controller-sdk-go/api"
+	"github.com/deis/controller-sdk-go/config"
+)
+
+// AnnotationList lists an app annotations
+func (d *DeisCmd) AnnotationList(appID string, format string) error {
+	settings, appID, err := load(d.ConfigFile, appID)
+
+	if err != nil {
+		return err
+	}
+
+	config, err := config.List(settings.Client, appID)
+	if d.checkAPICompatibility(settings.Client, err) != nil {
+		return err
+	}
+	var configOutput = new(bytes.Buffer)
+
+	/*** type casting to map[string]interface{}
+	 * the sortKeys function takes a map[string]interface{}, but config.Annotations
+	 * is of type map[string]Annotations. Because of that type difference, the function
+	 * cannot take `config.Annotation` directly. So we have two main options, either
+	 * replicate the sortKeys function changing only its type signature, or copy
+	 * the map into another one with the correct type. This approach was selected
+	 * because this map will likely have 2 or 3 elements, so copying this way
+	 * should not impact negatively on performance and will keep the code cleaner.
+	 */
+	interfacedStuff := make(map[string]interface{})
+	for k, v := range config.Annotations {
+		interfacedStuff[k] = v
+	}
+
+	sortedTypes := sortKeys(interfacedStuff)
+
+	for _, appType := range sortedTypes {
+		annotations := config.Annotations[appType]
+		keys := sortKeys(annotations)
+		switch format {
+		case "oneline":
+			fmt.Fprintf(configOutput, "%s:", appType)
+			for _, key := range keys {
+				value := annotations[key]
+				fmt.Fprintf(configOutput, " %s=%s", key, value)
+			}
+			fmt.Fprintf(configOutput, "\n")
+		case "diff":
+			fmt.Fprintf(configOutput, "%s:\n", appType)
+			for _, key := range keys {
+				value := annotations[key]
+				fmt.Fprintf(configOutput, "    %s=%s\n", key, value)
+			}
+		default:
+			fmt.Fprintf(configOutput, "=== %s Annotations\n", appType)
+			prettyPrintAnnotations := make(map[string]string)
+            for _, key := range keys {
+				value := annotations[key]
+				prettyPrintAnnotations[key] = value.(string)
+			}
+			fmt.Fprint(configOutput, prettyprint.PrettyTabs(prettyPrintAnnotations, 6))
+		}
+	}
+
+	d.Print(configOutput)
+	return nil
+}
+
+// AnnotationSet sets an app's annotations.
+func (d *DeisCmd) AnnotationSet(appID string, appType string, annotationCommandLine []string) error {
+	settings, appID, err := load(d.ConfigFile, appID)
+
+	if err != nil {
+		return err
+	}
+
+	annotations, err := parseAnnotations(annotationCommandLine)
+	if err != nil {
+		return err
+	}
+
+	d.Print("Creating Annotations... ")
+
+	quit := progress(d.WOut)
+
+	appAnnotations := make(map[string]api.Annotation)
+	appAnnotations[appType] = annotations
+	configObj := api.Config{Annotations: appAnnotations}
+	configObj, err = config.Set(settings.Client, appID, configObj)
```

Comment:
config.Set vai setar apenas Annotations e deixar os outros intactos?

Reply:
Sim, ele seta só o que for existir no `configObj`, então ele só vai adicionar algo se tiver nessa struct